File: packages/protocol/src/transfer-receiver.ts
/**
 * Receiving half of the transport-agnostic transfer engine.
 *
 * Blocks are authenticated, hashed, and committed before acknowledgement. A valid block arriving
 * ahead of the next required index exposes a gap: it is deliberately not committed, and the
 * receiver requests the missing block. Duplicate retransmissions are reverified and acknowledged
 * without being written or hashed twice. Any AEAD or block-hash failure aborts immediately.
 */

import { open, seal, type FrameHeaderInput } from './aead.js';
import type { DirectionalKey } from './keyschedule.js';
import { FrameType, type ControlOp, type FileEntry, type Manifest } from './transfer.js';
import { DEFAULT_INFLIGHT_BLOCKS, FRAME_VERSION } from './constants.js';
import { sha256 } from './webcrypto.js';
import { bytesToHex } from './bytes.js';
import {
  TransferError,
  singleSinkDestination,
  type Destination,
  type Digest,
  type Sink,
} from './transfer-ports.js';
import { decodeControl, encodeControl } from './transfer-messages.js';
import type { TransferRunState } from './transfer-sender.js';
import { validateManifest } from './safe-path.js';
import { completionDigest } from './transfer-set.js';

export interface ReceiveResult {
  files: FileEntry[];
  digests: string[];
  totalSize: number;
  /** Compatibility aliases for the first file and transfer completion digest. */
  file: FileEntry;
  digest: string;
}

export interface TransferReceiverOptions {
  send(frame: Uint8Array): void | Promise<void>;
  sendDir: DirectionalKey;
  recvDir: DirectionalKey;
  sendCounterStart: number;
  recvCounterStart: number;
  createDigest(): Digest;
  /** One-file compatibility destination. Exactly one of sink/destination is required. */
  sink?: Sink;
  destination?: Destination;
  /** Reports bytes only after verify-and-sink. */
  onProgress?(acknowledgedBytes: number): void;
  onFileProgress?(fileIdx: number, fileBytes: number, acknowledgedBytes: number): void;
  onStateChange?(state: TransferRunState): void;
  /** Called once with the validated complete file set before any destination opens. */
  onManifestSet?(manifest: Manifest): void | Promise<void>;
  /** Called for each file immediately before its sink opens. */
  onManifest?(file: FileEntry): void | Promise<void>;
}

export class TransferReceiver {
  private readonly o: TransferReceiverOptions;
  private readonly destination: Destination;

  private sendCounter: number;
  private recvCounter: number;
  private manifest: Manifest | undefined;
  private fileIdx = 0;
  private file: FileEntry | undefined;
  private sink: Sink | undefined;
  private digest: Digest | undefined;
  private readonly digests: string[] = [];
  private acknowledged = 0;
  private nextBlock = 0;
  private assemblingBlock = -1;
  private assemblingFileIdx = -1;
  private blockBuf: Uint8Array | undefined;
  private blockReceived = 0;
  private readonly seenAhead = new Set<number>();
  private nackOutstanding: number | undefined;
  private paused = false;

  private resolveDone!: (r: ReceiveResult) => void;
  private rejectDone!: (e: Error) => void;
  readonly done: Promise<ReceiveResult>;
  private inbound: Promise<void> = Promise.resolve();
  private outbound: Promise<void> = Promise.resolve();
  private settled = false;

  constructor(opts: TransferReceiverOptions) {
    this.o = opts;
    if ((opts.sink === undefined) === (opts.destination === undefined)) {
      throw new Error('exactly one of sink or destination is required');
    }
    this.destination = opts.destination ?? singleSinkDestination(opts.sink!);
    this.sendCounter = opts.sendCounterStart;
    this.recvCounter = opts.recvCounterStart;
    this.done = new Promise<ReceiveResult>((res, rej) => {
      this.resolveDone = res;
      this.rejectDone = rej;
    });
    this.done.catch(() => {});
  }

  /** Feed one inbound encrypted frame from the sender. */
  handle(frame: Uint8Array): void {
    this.inbound = this.inbound
      .then(() => this.process(frame))
      .catch((e: unknown) => {
        void this.abortWith(
          e instanceof TransferError
            ? e
            : new TransferError('integrity', e instanceof Error ? e.message : String(e)),
        );
      });
  }

  /** Ask the sender to stop producing data frames. Buffered transport bytes may still drain. */
  pause(): void {
    if (this.settled || this.paused) return;
    this.setPaused(true);
    void this.sendControl(FrameType.Control, { type: FrameType.Control, op: 'pause' }).catch(
      (e: unknown) => void this.abortWith(asIntegrityError(e)),
    );
  }

  /** Ask the sender to continue. */
  resume(): void {
    if (this.settled || !this.paused) return;
    this.setPaused(false);
    void this.sendControl(FrameType.Control, { type: FrameType.Control, op: 'resume' }).catch(
      (e: unknown) => void this.abortWith(asIntegrityError(e)),
    );
  }

  /** Best-effort peer notification followed by terminal local cancellation. */
  cancel(reason = 'canceled'): void {
    if (this.settled) return;
    this.o.onStateChange?.('canceled');
    void this.sendControl(FrameType.Control, { type: FrameType.Control, op: 'cancel' })
      .catch(() => {})
      .finally(
        () => void this.abortWith(new TransferError('canceled', reason), { notifyPeer: false }),
      );
  }

  private async process(frame: Uint8Array): Promise<void> {
    if (this.settled) return;
    let opened;
    try {
      opened = await open(this.o.recvDir, this.recvCounter++, frame);
    } catch (e) {
      throw new TransferError(
        'integrity',
        e instanceof Error ? e.message : 'unable to authenticate transfer frame',
      );
    }
    switch (opened.header.type) {
      case FrameType.Manifest:
        return this.applyManifest(opened.plaintext);
      case FrameType.BlockData:
        return this.onBlockData(
          opened.header.fileIdx,
          opened.header.blockIdx,
          opened.header.frameOff,
          opened.plaintext,
        );
      case FrameType.BlockHash:
        return this.onBlockHash(opened.plaintext);
      case FrameType.Control:
        return this.onControl(opened.plaintext);
      case FrameType.Complete:
        return this.onComplete(opened.plaintext);
      case FrameType.Fail:
        return this.onPeerFail(opened.plaintext);
      default:
        throw new TransferError(
          'integrity',
          `unexpected receiver-inbound type ${opened.header.type}`,
        );
    }
  }

  private async applyManifest(payload: Uint8Array): Promise<void> {
    if (this.manifest) throw new TransferError('integrity', 'duplicate manifest');
    const msg = decodeControl(payload);
    if (msg.type !== FrameType.Manifest) throw new TransferError('integrity', 'expected manifest');
    let manifest;
    try {
      manifest = validateManifest(msg);
    } catch (e) {
      throw new TransferError('integrity', e instanceof Error ? e.message : String(e));
    }
    try {
      await this.destination.prepare(manifest);
      await this.o.onManifestSet?.(manifest);
    } catch (e) {
      throw e instanceof TransferError
        ? e
        : new TransferError('sink_error', e instanceof Error ? e.message : String(e));
    }
    this.manifest = manifest;
    await this.openNextFile();
  }

  /** Open the next file and immediately verify/close consecutive empty files. */
  private async openNextFile(): Promise<void> {
    const manifest = this.manifest;
    if (!manifest) throw new TransferError('integrity', 'file opened before manifest');
    while (this.fileIdx < manifest.files.length) {
      const file = manifest.files[this.fileIdx]!;
      this.file = file;
      this.nextBlock = 0;
      this.seenAhead.clear();
      this.nackOutstanding = undefined;
      this.digest = this.o.createDigest();
      try {
        await this.o.onManifest?.(file);
        this.sink = await this.destination.open(file);
      } catch (e) {
        throw e instanceof TransferError
          ? e
          : new TransferError('sink_error', e instanceof Error ? e.message : String(e));
      }
      if (file.blocks > 0) return;
      await this.finishCurrentFile();
    }
    this.file = undefined;
    this.sink = undefined;
    this.digest = undefined;
  }

  private async finishCurrentFile(): Promise<void> {
    const file = this.file;
    const digest = this.digest;
    const sink = this.sink;
    if (!file || !digest || !sink) throw new TransferError('integrity', 'no active file');
    const got = await digest.hexDigest();
    if (got !== file.fileDigest) {
      throw new TransferError('digest_mismatch', `file ${file.idx} digest mismatch`);
    }
    try {
      await sink.close();
    } catch (e) {
      throw e instanceof TransferError
        ? e
        : new TransferError('sink_error', e instanceof Error ? e.message : String(e));
    }
    this.digests[file.idx] = got;
    this.o.onFileProgress?.(file.idx, file.size, this.acknowledged);
    this.fileIdx++;
    this.file = undefined;
    this.sink = undefined;
    this.digest = undefined;
  }

  private onBlockData(
    fileIdx: number,
    blockIdx: number,
    frameOff: number,
    payload: Uint8Array,
  ): void {
    const manifest = this.manifest;
    if (!manifest) throw new TransferError('integrity', 'block_data before manifest');
    const entry = manifest.files[fileIdx];
    if (!entry || fileIdx > this.fileIdx || blockIdx < 0 || blockIdx >= entry.blocks) {
      throw new TransferError('integrity', `block_data outside manifest: ${blockIdx}`);
    }
    if (frameOff === 0) {
      if (this.blockBuf) throw new TransferError('integrity', 'new block before block_hash');
      const blockLen = Math.min(entry.blockSize, entry.size - blockIdx * entry.blockSize);
      this.assemblingFileIdx = fileIdx;
      this.assemblingBlock = blockIdx;
      this.blockBuf = new Uint8Array(blockLen);
      this.blockReceived = 0;
    }
    if (!this.blockBuf || this.assemblingFileIdx !== fileIdx || this.assemblingBlock !== blockIdx) {
      throw new TransferError('integrity', `unexpected block fragment ${blockIdx}`);
    }
    if (frameOff !== this.blockReceived || frameOff + payload.length > this.blockBuf.length) {
      throw new TransferError('integrity', `invalid frame offset in block ${blockIdx}`);
    }
    this.blockBuf.set(payload, frameOff);
    this.blockReceived += payload.length;
  }

  private async onBlockHash(payload: Uint8Array): Promise<void> {
    const block = this.blockBuf;
    if (!this.manifest || !block) {
      throw new TransferError('integrity', 'block_hash without a block');
    }
    const msg = decodeControl(payload);
    if (msg.type !== FrameType.BlockHash) {
      throw new TransferError('integrity', 'expected block_hash');
    }
    if (msg.fileIdx !== this.assemblingFileIdx || msg.blockIdx !== this.assemblingBlock) {
      throw new TransferError('integrity', 'block_hash does not match assembled block');
    }
    if (this.blockReceived !== block.length) throw new TransferError('integrity', 'short block');
    const got = bytesToHex(await sha256(block));
    if (got !== msg.sha256) {
      throw new TransferError('integrity', `block ${msg.blockIdx} hash mismatch`);
    }

    this.blockBuf = undefined;
    this.blockReceived = 0;
    this.assemblingBlock = -1;
    this.assemblingFileIdx = -1;

    if (msg.fileIdx < this.fileIdx) {
      await this.sendAck(msg.fileIdx, msg.blockIdx);
      return;
    }
    const file = this.file;
    const sink = this.sink;
    const digest = this.digest;
    if (!file || !sink || !digest || msg.fileIdx !== this.fileIdx) {
      throw new TransferError('integrity', 'block_hash for inactive file');
    }

    if (msg.blockIdx < this.nextBlock) {
      await this.sendAck(msg.fileIdx, msg.blockIdx);
      return;
    }
    if (msg.blockIdx > this.nextBlock) {
      if (this.seenAhead.size < DEFAULT_INFLIGHT_BLOCKS) this.seenAhead.add(msg.blockIdx);
      await this.requestMissing();
      return;
    }

    const offset = this.nextBlock * file.blockSize;
    try {
      await sink.write(offset, block);
    } catch (e) {
      throw e instanceof TransferError
        ? e
        : new TransferError('sink_error', e instanceof Error ? e.message : String(e));
    }
    digest.update(block);
    this.nextBlock++;
    this.nackOutstanding = undefined;
    this.acknowledged += block.length;
    this.o.onProgress?.(this.acknowledged);
    this.o.onFileProgress?.(file.idx, offset + block.length, this.acknowledged);
    await this.sendAck(msg.fileIdx, msg.blockIdx);

    if (this.nextBlock === file.blocks) {
      await this.finishCurrentFile();
      await this.openNextFile();
    } else if (this.seenAhead.delete(this.nextBlock)) {
      await this.requestMissing();
    }
  }

  private async sendAck(fileIdx: number, blockIdx: number): Promise<void> {
    await this.sendControl(FrameType.Ack, {
      type: FrameType.Ack,
      fileIdx,
      blockIdx,
    });
  }

  private async requestMissing(): Promise<void> {
    if (this.nackOutstanding === this.nextBlock) return;
    this.nackOutstanding = this.nextBlock;
    await this.sendControl(FrameType.Nack, {
      type: FrameType.Nack,
      fileIdx: this.fileIdx,
      blockIdx: this.nextBlock,
      reason: 'missing',
    });
  }

  private onControl(payload: Uint8Array): void {
    const msg = decodeControl(payload);
    if (msg.type !== FrameType.Control) throw new TransferError('integrity', 'expected control');
    this.applyRemoteControl(msg.op);
  }

  private applyRemoteControl(op: ControlOp): void {
    switch (op) {
      case 'pause':
        this.setPaused(true);
        return;
      case 'resume':
        this.setPaused(false);
        return;
      case 'cancel':
        this.o.onStateChange?.('canceled');
        void this.abortWith(new TransferError('canceled', 'peer canceled the transfer'), {
          notifyPeer: false,
        });
    }
  }

  private setPaused(paused: boolean): void {
    if (this.paused === paused || this.settled) return;
    this.paused = paused;
    this.o.onStateChange?.(paused ? 'paused' : 'running');
  }

  private async onPeerFail(payload: Uint8Array): Promise<void> {
    const msg = decodeControl(payload);
    if (msg.type !== FrameType.Fail) throw new TransferError('integrity', 'expected fail');
    await this.abortWith(new TransferError(msg.reason, `sender failed: ${msg.reason}`), {
      notifyPeer: false,
    });
  }

  private async onComplete(payload: Uint8Array): Promise<void> {
    const manifest = this.manifest;
    if (!manifest) throw new TransferError('integrity', 'complete before manifest');
    const msg = decodeControl(payload);
    if (msg.type !== FrameType.Complete) throw new TransferError('integrity', 'expected complete');
    if (this.file && this.nextBlock !== this.file.blocks) {
      await this.requestMissing();
      return;
    }
    if (this.fileIdx !== manifest.files.length) {
      throw new TransferError('integrity', 'complete before every file was received');
    }
    const got = await completionDigest(manifest.files);
    if (got !== msg.fileDigest) throw new TransferError('digest_mismatch', 'file-set mismatch');
    try {
      await this.destination.close();
    } catch (e) {
      throw e instanceof TransferError
        ? e
        : new TransferError('sink_error', e instanceof Error ? e.message : String(e));
    }
    await this.sendControl(FrameType.Done, { type: FrameType.Done });
    this.settled = true;
    this.resolveDone({
      files: manifest.files,
      digests: [...this.digests],
      totalSize: manifest.totalSize,
      file: manifest.files[0]!,
      digest: got,
    });
  }

  private async abortWith(
    err: TransferError,
    options: { notifyPeer?: boolean } = {},
  ): Promise<void> {
    if (this.settled) return;
    this.settled = true;
    if (options.notifyPeer !== false) {
      try {
        await this.sendControl(FrameType.Fail, { type: FrameType.Fail, reason: err.reason });
      } catch {
        // The channel may already be unavailable.
      }
    }
    try {
      await this.destination.abort(err.reason);
    } catch {
      // Sink abort is best-effort after the first terminal failure.
    }
    this.rejectDone(err);
  }

  private async sendControl(
    type: FrameType,
    msg: Parameters<typeof encodeControl>[0],
  ): Promise<void> {
    const header: FrameHeaderInput = {
      version: FRAME_VERSION,
      type,
      flags: 0,
      fileIdx: 0,
      blockIdx: 0,
      frameOff: 0,
    };
    await this.sendFrame(header, encodeControl(msg));
  }

  /** Serialize sealing so UI controls and acknowledgements cannot reuse a nonce. */
  private sendFrame(header: FrameHeaderInput, payload: Uint8Array): Promise<void> {
    const task = this.outbound.then(async () => {
      const frame = await seal(this.o.sendDir, this.sendCounter++, header, payload);
      await this.o.send(frame);
    });
    this.outbound = task.catch(() => {});
    return task;
  }
}

function asIntegrityError(e: unknown): TransferError {
  return e instanceof TransferError
    ? e
    : new TransferError('integrity', e instanceof Error ? e.message : String(e));
}

File: packages/wire/transfer_receiver.go
package wire

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"sync"
)

// Receiver is the receiving half of the transfer engine (design §5, §6). Frames arrive over an
// ordered, reliable channel, so blocks assemble one at a time, in order. For each block the
// receiver verifies its SHA-256 (from block_hash) BEFORE writing to the sink, feeds the verified
// bytes into the whole-file streaming digest, then signals block_recv (the in-flight window). On
// complete it compares the recomputed digest to the sender's and replies done, or
// fail{digest_mismatch}. A GCM open failure or any hash mismatch yields fail{integrity} and an
// abort — never a silent retry (design §5.5). It is the Go twin of
// packages/protocol/src/transfer-receiver.ts.
//
// Handle is called from the transport read loop and MUST be called serially — the ordered
// DataChannel satisfies this, and each frame consumes the next receive counter. Wait blocks a
// separate goroutine until the transfer settles.

// ReceiveResult is the outcome of a successful receive: the manifest entry and the verified
// whole-file digest (hex).
type ReceiveResult struct {
	File   FileEntry
	Digest string
}

// ReceiverOptions configures a Receiver.
type ReceiverOptions struct {
	Send             func(frame []byte) error
	SendDir          DirectionalKey
	RecvDir          DirectionalKey
	SendCounterStart uint64
	RecvCounterStart uint64
	CreateDigest     func() Digest // nil selects a streaming SHA-256 digest
	Sink             Sink
	OnProgress       func(receivedBytes int64)
	// OnManifest is called once when the manifest decodes, before any block is written — it lets
	// the host open a destination named after the file without the engine knowing about it. A
	// non-nil error aborts the transfer.
	OnManifest func(file FileEntry) error
}

// Receiver drives one file receive. Construct it with NewReceiver.
type Receiver struct {
	o      ReceiverOptions
	digest Digest

	// Counters and assembly state are touched only by Handle's (serial) goroutine.
	sendCounter uint64
	recvCounter uint64
	file        *FileEntry
	curBlock    int
	blockBuf    []byte
	blockLen    int
	received    int

	mu      sync.Mutex
	done    chan struct{}
	settled bool
	result  ReceiveResult
	err     error
}

// NewReceiver builds a Receiver, defaulting to a streaming SHA-256 digest when none is given.
func NewReceiver(opts ReceiverOptions) *Receiver {
	if opts.CreateDigest == nil {
		opts.CreateDigest = NewSHA256Digest
	}
	return &Receiver{
		o:           opts,
		digest:      opts.CreateDigest(),
		sendCounter: opts.SendCounterStart,
		recvCounter: opts.RecvCounterStart,
		done:        make(chan struct{}),
	}
}

// Wait blocks until the transfer settles, returning the result on done or the error on fail.
// Canceling ctx aborts the receive with FailCanceled in bounded time, so a dropped peer never
// parks Wait forever.
func (r *Receiver) Wait(ctx context.Context) (ReceiveResult, error) {
	select {
	case <-r.done:
	case <-ctx.Done():
		// abortWith is idempotent and closes r.done; if the receive already settled it is a
		// no-op and the <-r.done below returns immediately with the real result.
		r.abortWith(NewTransferError(FailCanceled, ctx.Err().Error()))
		<-r.done
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.result, r.err
}

// Handle feeds one inbound frame from the sender (manifest / block_data / block_hash / complete).
// Any error aborts the transfer with the appropriate fail reason (integrity unless the error
// already carries one).
func (r *Receiver) Handle(frame []byte) {
	if err := r.process(frame); err != nil {
		var te *TransferError
		if !errors.As(err, &te) {
			te = NewTransferError(FailIntegrity, err.Error())
		}
		r.abortWith(te)
	}
}

// --- internal ---------------------------------------------------------------

func (r *Receiver) process(frame []byte) error {
	if r.isSettled() {
		return nil
	}
	ctr := r.recvCounter
	r.recvCounter++
	opened, err := Open(r.o.RecvDir, ctr, frame) // open failure → integrity abort
	if err != nil {
		return NewTransferError(FailIntegrity, err.Error())
	}
	switch opened.Header.Type {
	case FrameManifest:
		return r.applyManifest(opened.Plaintext)
	case FrameBlockData:
		return r.onBlockData(opened.Header.BlockIdx, opened.Header.FrameOff, opened.Plaintext)
	case FrameBlockHash:
		return r.onBlockHash(opened.Plaintext)
	case FrameComplete:
		return r.onComplete(opened.Plaintext)
	default:
		return NewTransferError(FailIntegrity,
			fmt.Sprintf("unexpected receiver-inbound type %d", opened.Header.Type))
	}
}

func (r *Receiver) applyManifest(payload []byte) error {
	msg, err := DecodeControl(payload)
	if err != nil {
		return NewTransferError(FailIntegrity, err.Error())
	}
	m, ok := msg.(*Manifest)
	if !ok {
		return NewTransferError(FailIntegrity, "expected manifest")
	}
	if len(m.Files) == 0 {
		return NewTransferError(FailIntegrity, "manifest has no files")
	}
	f := m.Files[0]
	r.file = &f
	if r.o.OnManifest != nil {
		return r.o.OnManifest(f)
	}
	return nil
}

func (r *Receiver) onBlockData(blockIdx uint32, frameOff uint32, payload []byte) error {
	if r.file == nil {
		return NewTransferError(FailIntegrity, "block_data before manifest")
	}
	if int(blockIdx) != r.curBlock {
		return NewTransferError(FailIntegrity, fmt.Sprintf("out-of-order block %d", blockIdx))
	}
	if frameOff == 0 {
		// The final block is short: cap its buffer at the file's remaining bytes.
		r.blockLen = r.file.BlockSize
		if rem := r.file.Size - int64(blockIdx)*int64(r.file.BlockSize); int64(r.blockLen) > rem {
			r.blockLen = int(rem)
		}
		r.blockBuf = make([]byte, r.blockLen)
		r.received = 0
	}
	copy(r.blockBuf[frameOff:], payload)
	r.received += len(payload)
	return nil
}

func (r *Receiver) onBlockHash(payload []byte) error {
	if r.file == nil || r.blockBuf == nil {
		return NewTransferError(FailIntegrity, "block_hash without a block")
	}
	msg, err := DecodeControl(payload)
	if err != nil {
		return NewTransferError(FailIntegrity, err.Error())
	}
	bh, ok := msg.(*BlockHash)
	if !ok {
		return NewTransferError(FailIntegrity, "expected block_hash")
	}
	if r.received != r.blockLen {
		return NewTransferError(FailIntegrity, "short block")
	}
	sum := sha256.Sum256(r.blockBuf)
	if hex.EncodeToString(sum[:]) != bh.SHA256 {
		return NewTransferError(FailIntegrity, fmt.Sprintf("block %d hash mismatch", bh.BlockIdx))
	}

	// Verified: write, fold into the whole-file digest, then open the window one more block. A
	// sink error propagates as-is, so a sink returning a TransferError keeps its reason.
	offset := int64(r.curBlock) * int64(r.file.BlockSize)
	if err := r.o.Sink.Write(offset, r.blockBuf); err != nil {
		return err
	}
	r.digest.Update(r.blockBuf)
	if r.o.OnProgress != nil {
		r.o.OnProgress(offset + int64(r.blockLen))
	}
	if err := r.sendControl(NewBlockRecv(0, r.curBlock)); err != nil {
		return err
	}
	r.curBlock++
	r.blockBuf = nil
	return nil
}

func (r *Receiver) onComplete(payload []byte) error {
	if r.file == nil {
		return NewTransferError(FailIntegrity, "complete before manifest")
	}
	msg, err := DecodeControl(payload)
	if err != nil {
		return NewTransferError(FailIntegrity, err.Error())
	}
	c, ok := msg.(*Complete)
	if !ok {
		return NewTransferError(FailIntegrity, "expected complete")
	}
	got := r.digest.HexDigest()
	if got != c.FileDigest {
		return NewTransferError(FailDigestMismatch, "whole-file digest mismatch")
	}
	if err := r.o.Sink.Close(); err != nil {
		return err
	}
	if err := r.sendControl(NewDone()); err != nil {
		return err
	}
	r.settle(ReceiveResult{File: *r.file, Digest: got})
	return nil
}

// abortWith settles the transfer as failed and, best-effort, tells the peer and the sink. Only
// the first abort (or a prior settle) takes effect.
func (r *Receiver) abortWith(err *TransferError) {
	r.mu.Lock()
	if r.settled {
		r.mu.Unlock()
		return
	}
	r.settled = true
	r.err = err
	r.mu.Unlock()

	_ = r.sendControl(NewFail(err.Reason)) // best-effort — the channel may already be gone
	_ = r.o.Sink.Abort(string(err.Reason)) // best-effort
	close(r.done)
}

// settle records a successful finish; only the first settle (or a prior abort) takes effect.
func (r *Receiver) settle(res ReceiveResult) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.settled {
		return
	}
	r.settled = true
	r.result = res
	close(r.done)
}

func (r *Receiver) isSettled() bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.settled
}

// sendControl seals and sends a control message under the next send counter.
func (r *Receiver) sendControl(msg ControlMsg) error {
	payload, err := EncodeControl(msg)
	if err != nil {
		return err
	}
	frame, err := Seal(r.o.SendDir, r.sendCounter, FrameHeaderInput{
		Version: FrameVersion, Type: msg.FrameType(),
	}, payload)
	if err != nil {
		return err
	}
	r.sendCounter++
	return r.o.Send(frame)
}
